metsim/metsim.py

```diff
@@ -236,119 +281,61 @@ def ascii_out_preprocess(self):
-        self.output = Dataset(out_file, 'w')
-
-        # Number of timesteps
-        disagg = int(MetSim.params['time_step']) < cnst.MIN_PER_DAY
-        if disagg:
-            delta = pd.Timedelta('1 days')
-        else:
-            delta = pd.Timedelta('0 days')
-        n_ts = len(pd.date_range(MetSim.params['start'],
-                   MetSim.params['stop'] + delta,
-                   freq="{}T".format(MetSim.params['time_step'])))
-
-        # Create dimensions
-        self.output.createDimension('lat', self.domain_shape[0])
-        self.output.createDimension('lon', self.domain_shape[1])
-        self.output.createDimension('time', n_ts)
-
-        # Create output variables
-        for varname in self.domain_variables:
-            da = getattr(self, varname)
-            var = self.output.createVariable(varname, da.dtype, da.dims)
-            var[:] = da.values
-            # Copy the attributes
-            for key, val in da.attrs.items():
-                var.setncattr(key, val)
+        self.output_filename = os.path.join(self.params['out_dir'], "forcing.nc")
 
-        for varname in MetSim.params['out_vars']:
-            var = self.output.createVariable(varname, 'd',
-                                             ('time', 'lat', 'lon'))
-            # set each variables attributes
-            for key, val in attrs.get(varname, {}).items():
-                var.setncattr(key, val)
-
-        # set global attributes
-        for key, val in attrs.get('_global', {}).items():
-            self.output.setncattr(key, val)
-
-    def create_io_process(self, queue):
-        """Launch the IO process"""
-        out_preprocess = {"ascii": self.ascii_out_preprocess,
-                          "netcdf": self.netcdf_out_preprocess}
-        out_preprocess[MetSim.params['out_format']]()
-        # Wait for data to come in
-        while 1:
-            data = queue.get()
-            if data == "done":
-                break
-            self.write(data)
-
-        # Close the dataset when done
-        if self.output:
-            self.output.close()
-
-    def write(self, data: dict):
+    def write(self):
         """
         Dispatch to the right function based on the configuration given
         """
         dispatch = {
                 'netcdf': self.write_netcdf,
                 'ascii': self.write_ascii
                 }
-        dispatch[MetSim.params.get('out_format', 'netcdf').lower()](data)
+        dispatch[MetSim.params.get('out_format', 'netcdf').lower()]()
 
-    def write_netcdf(self, data: dict):
+    def write_netcdf(self):
         """Write out as NetCDF to the output file"""
         print("Writing netcdf...")
-        for (i, j), df in data.items():
-            for varname in MetSim.params['out_vars']:
-                self.output.variables[varname][:, i, j] = df[varname].values
+        self.output.to_netcdf(self.output_filename,
+                              unlimited_dims=['time'],
```

Comment:
Rather than hard-coding the strings (in many of these dicts) , it would be preferable to use variables that would then have values like `time` or `elev`, but which could be changed in one place in the code. It would make it easier to maintain.